### DISCUSSION TITLE
Clothesline Method with Quick Add

### DISCUSSION THREAD

**Discussion by zsviczian** [zsviczian/obsidian-excalidraw-plugin#138] [Show and tell]
![Intro Slide YouTube - QuickAdd](https://user-images.githubusercontent.com/14358394/132582992-10870f24-ee27-49c4-83b8-3414239a9620.png)

See this 📽[video](https://youtu.be/b6jhQrRMTvs) for a walkthrough of this example.

# Plugins used in the example (beyond Excalidraw of course)
### CodeView
I use my [CodeView plugin](https://github.com/zsviczian/obsidian-codeeditor/releases/tag/0.0.3) for editing javascript files within Obsidian. This plugin is not yet available in community plugins, but you can install it following the link. Also, you can use an external editor to edit javascript files (e.g. notepad++).

### QuickAdd
[QuickAdd](https://github.com/chhoumann/quickadd) is available in community plugins.

# Clothesline template drawing
Download this file and add it to your vault. This is the empty clothesline drawing
[Empty Clothesline.md](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/7131551/Empty.Clothesline.md)

# The scripts
QuickAdd reads scripts from the "scripts" folder of your Vault. Create a "scripts" folder and then two .js files in it. Copy the contents of these files into the two files respectively. You may need to use CTRL+SHIFT+V to paste to preserve formatting.

### clothesline-step-1.js
```javascript
module.exports = async (params) => {
  const {quickAddApi: {inputPrompt}} = params;
  params.variables["cardTitle"] = await inputPrompt("Title of the new card");

  const path = app.workspace.activeLeaf.view.file.path;
  const lastDot = path.lastIndexOf(".");
  const length = (lastDot === -1) ? path.length : lastDot;
  params.variables["cardFilePath"] = `${path.substr(0,length)} Card Details`;
}
```

### clothesline-step-2.js
```javascript
module.exports = (params) => {
	//Initialize ExcalidrawAutomate
	const ea = ExcalidrawAutomate;
	ea.reset();
	ea.style.fontSize = 9;
	ea.style.textAlign = "center";
	ea.setFontFamily(1);
	ea.setFillStyle(2); //solid fill
	ea.style.strokeColor = "black";
	ea.setView("active");
	
	//Card
	const cardWidth = 76;
	const cardHeight = 65;
	//check if previous card is selected, if yes, calculate top-left position of new card
	const selectedElement = ea.getViewSelectedElement();
	let [offsetX,offsetY] = [0,0];
	if(selectedElement) {
		offsetX = selectedElement.x;
		offsetY = selectedElement.y+cardHeight+5;
	}
	ea.style.backgroundColor = "WhiteSmoke";
	const idCard = ea.addRect(offsetX,offsetY,cardWidth,cardHeight);
	
	//Clip
	const clipWidth = 8;
	const colors = ["#40c057","#7950f2","#228be6","#fa5252","#fab005","#868e96"]
	const clipcolor = colors[Math.floor(Math.random()*colors.length)];
	ea.style.backgroundColor = clipcolor;
	const idClip  = ea.addRect(offsetX+(cardWidth-clipWidth)/2,offsetY-9,clipWidth,14);
	const idClipSpring = ea.addLine(
		[[offsetX+(cardWidth-clipWidth)/2,offsetY-2],
		 [offsetX+(cardWidth+clipWidth)/2,offsetY-2]]
	);

	//Message
	const cardTitle = params.variables["cardTitle"];
	const cardTitleWrapped = ea.wrapText(cardTitle,15);
	const idText = ea.addText(offsetX,offsetY+10,cardTitleWrapped);
	const textElement = ea.getElement(idText);
	textElement.x =  offsetX+(cardWidth - textElement.width)/2; //center text element on card

	//Link to details
	const cardDetailsFile = app.vault.getAbstractFileByPath(`${params.variables["cardFilePath"]}.md`);
	const fileName = app.metadataCache.fileToLinktext(cardDetailsFile,app.workspace.activeLeaf.view.file.path,true);
	const idDetails = ea.addText(offsetX+cardWidth-11,offsetY+cardHeight-11,`[[${fileName}#${cardTitle}|»]]`);
	const textDetails = ea.getElement(idDetails);
	textDetails.width = 10; //In preview mode the text will be much shorter than the markdown supplied

	//Add to the view
	ea.addToGroup([idCard,idClip,idClipSpring,idText,idDetails]);
	ea.addElementsToView(!selectedElement,true); //recalculate position around pointer if no element is selected, save file after action
}
```

**Comment by erex** [comment]
@zsviczian trying to duplicate your video steps (without knowing much about QuickAdd, Javascript or Templater).

Not sufficiently patient to watch all of the explanatory QuickAdd video, I was first thrown off by not saving the scripts above **outside** of Obsidian (which will append a .md extention).  Additionally, I needed to figure out the location of the saved scripts (in folder specified, not in QuickAdd, but in Templater).

![image](https://user-images.githubusercontent.com/330744/132680372-bd8bdd39-3707-4a18-8f0f-65cf0fd700f7.png)


Once those issues were sorted and the QuickAdd macro was running, I encountered a new problem:
- cards are created in diagram as expected
- note containing titles are created
- but links are malformed, resulting in error message *unable to find section XXX in yyy card details"
  - I wonder if that is because the line in the second macro near the bottom defining `idDetails` is missing a space after the hash?

Troubles aside, I'm still enjoying the learning process

**Reply by erex** [reply]
Further experimentation after the work day
- problem does not reside with the `idDetails` line of code
- for some reason, the `Card Details` note creates a task list when I execute my `ctrl-1` macro resulting in 
- 
![image](https://user-images.githubusercontent.com/330744/132736554-df1bb8e5-e093-4ee2-bb2e-deaf38781f4a.png)

removing everything in front of the hash results in working transclusion

I don't know how I managed to have the macro create a task list

**Reply by zsviczian** [reply]
I can fully see how quickAdd, javascript and all can feel overwhelming at first. I am actively thinking about a solution where you can "install" macros from a community store into your Obsidian Excalidraw.
Considering your opening statement in the previous message, you have done great! 🥇

I hope you will be able to put your newly learned skills to good use! QuickAdd is a very powerful automation tool, which has many use-cases (e.g. I use it for interstitial journaling, and for creating future Note to Self reminders in future daily notes.

I will also be publishing my step-by-step guide for the "Prezi" automation with QuickAdd. The end solution is very slick... but you will need all the skills you learned with the Clothesline method to implement the prezi script.

Based on your description I assume you have ticked this box in the capture step in QuickAdd:
![image](https://user-images.githubusercontent.com/14358394/132748342-07b01e93-2ee7-4fc9-bfb2-11e51d2e0574.png)

**Reply by erex** [reply]
You are exactly right that I had "task" enabled.  Turning that off makes the macro perform as expected; thanks.

I'm using Excalidraw in teaching to demonstrate the relationship between components of a course I am teaching to the students.  I'm sure I'll find additional uses.  I can imagine a second (and third) prompt asking for the shape and colour of the card, permitting rudimentary flow chart symbols.

I presume the size of the transclusion window cannot be altered as it is not created by Excalidraw.  I ask because the transclusion window is a bit large for the usually small amount of information it contains.

Thanks again for your help.

**Reply by zsviczian** [reply]
Great! I am happy to hear that.
You could add additional prompts for additional shapes and colors, or you could create separate shortcuts for that. I encourage you to experiment. 

The size of the transclusion window comes from Obsidian, but I am sure there is a way to change its size. I may look at that sometime, though it is not a priority for me at the moment.
